llvm/lib/Target/AMDGPU/SIInstructions.td

```diff
@@ -2425,6 +2425,13 @@ def : GCNPat <(i1 imm:$imm),
   let WaveSizePredicate = isWave32;
 }
 
+let True16Predicate = UseRealTrue16Insts in
+foreach vt = [i32, v2i16] in
+def : GCNPat <
+  (vt (DivergentBinFrag<srl> VGPR_32:$src, (i32 16))),
+  (REG_SEQUENCE VGPR_32, (i16 (EXTRACT_SUBREG $src, hi16)), lo16, (V_MOV_B16_t16_e64 0, (i16 0x0000), 0), hi16)
```

Comment:
Probably should teach performSRLCombine for the 16-bit case, it already handles 64-to-32 in the same way and could be generalized

Reply:
Hi Matt. Thanks for the review. I checked the performSrlCombine function. I think it's currently checking for 64bit SRL with a dynamic right shift bits, which probably cannot be put in a isel pattern.

I think for this case since the right shift bit is a constant 16, it's probably better to directly place it in the isel pattern. (If I understand correctly we prefer isel pattern better than customization in the pass)

Reply:
I see. I will create another patch to address this.

Reply:
Hi Matt. I tried put this lowering into SrlCombine, but seems not working. The sdagcombine happens before ISel selection, and thus we don't have info about register at that time. Putting it in SrlCombine might get it lowered to sgpr_hi/lo16 which does not work.

Reply:
But the registers shouldn't matter? The problem isn't any different than the current 16-bit vs. 32-bit uniform problem. As a proxy you can just check if the value is uniform?

Reply:
Hi Matt. I think we are talking about two issues here.

The first thing is the 16-bit vs. 32-bit uniform problem, that we can teach the SrlCombine and the promoteUniformOp to use 16bit. Eventually they get lower to smaller size SRL, and finally get lowered to LSHRREV.

Another thing is the what this patch is trying to do. For VGPR 16bit right shift, we do not lower to LSHRREV, but use REG_SEQUENCE with .l/.h access. This depends on register type since we don't have sgpr 16 support yet